edx_lint/cmd/amnesty.py

```diff
@@ -0,0 +1,116 @@
+"""
+Commandline scripts to give one-time amnesty for specify linting error conditions.
+"""
+
+from __future__ import print_function
+
+from collections import namedtuple, defaultdict
+import re
+import sys
+
+import click
+
+PYLINT_PARSEABLE_REGEX = re.compile(
+    r"""^(?P<filename>[^:]+):(?P<linenum>\d+): """
+    r"""\[(?P<error_code>[^(]+)\((?P<error_name>[^)]+)\), (?P<function>[^\]]*)\] """
+    r"""(?P<error_msg>.*)"""
+)
+PYLINT_EXCEPTION_REGEX = re.compile(r"""\s*#\s*pylint:\s+disable=(?P<disables>[^#$]+?)(?=\s*(#|$))""")
+
+PylintError = namedtuple('PylintError', ['filename', 'linenum', 'error_code', 'error_name', 'function', 'error_msg'])
+
+def parse_pylint_output(pylint_output):
+    """
+    Parse the pylint output-format=parseable lines into PylintError tuples.
+    """
+    for line in pylint_output:
+        if not line.strip() or line.startswith("****"):
+            continue
+
+        parsed = PYLINT_PARSEABLE_REGEX.search(line)
+        if parsed is None:
+            raise Exception(
+                "Unable to parse {!r}. Please run pylint with the "
+                "--output-format=parseable option".format(
+                    line
+                )
+            )
+
+        parsed_dict = parsed.groupdict()
+        parsed_dict['linenum'] = int(parsed_dict['linenum'])
+        yield PylintError(**parsed_dict)
+
+
+def format_pylint_disables(error_names, tag=True):
+    """
+    Format a list of error_names into a 'pylint: disable=' line.
+    """
+    tag_str = "  # lint-amnesty" if tag else ""
+    if error_names:
+        return "  # pylint: disable={}{}".format(
+            ", ".join(sorted(error_names)),
+            tag_str,
+        )
+    else:
+        return ""
+
+
+def fix_pylint(line, errors):
+    """
+    Yield any modified versions of ``line`` needed to address the errors in ``errors``.
+    """
+    if not errors:
+        yield line
+        return
+
+    current = PYLINT_EXCEPTION_REGEX.search(line)
+    if current:
+        original_errors = {disable.strip() for disable in current.group('disables').split(',')}
+    else:
+        original_errors = set()
+
+    disabled_errors = set(original_errors)
+
+    for error in errors:
+        if error.error_name == 'useless-suppression':
+            parsed = re.search("""Useless suppression of '(?P<error_name>[^']+)'""", error.error_msg)
```

Comment:
Mostly, I pre-compiled them to make it easier to use them more than once. The pre-compilation isn't actually required for any kind of performance impact.